### ISSUE TITLE
crash report: bookmarking peaks with large numbers of samples

### ISSUE THREAD

**Issue by chubukov** [ElucidataInc/ElMaven#38]
In a real road test, I tried to use maven to do manual integration with about 100 samples loaded. I consistently got it to crash after doing 4-5 peaks, and there didn't seem to be any pattern as to what compounds it crashed on. I reproduced the same thing with the debug version, and it was always crashing on this line with a segmentation fault:

`        inline void addChild(const PeakGroup& child) { children.push_back(child); children.back().parent = this;   }`
(line 224 in `peakGroup.h`)

I can't figure out why it's segfaulting. `this` looks like a valid peakGroup and `children` is not null (it always has size=1 at the time of crash).

Could multithreading be involved somehow?

**Comment by GeorgeSabu** [comment]
@chubukov I tried loading 116 samples and tried the same. But I was not able to reproduce this. Could you please share with me  to the dataset and compound database with which you got this issue?

**Comment by chubukov** [comment]
@GeorgeSabu Uploaded my dataset to dropbox along with a couple of videos.

A couple of notes:
- I think it does have something to do with the data -- it seems to crash more often on crappy peaks. I haven't figured out if it's completely deterministic or not.
- Previously I was also going too fast on occasion and getting the behavior from #37 where the peak would not get the isotopes in the peak table. That seemed to exacerbate the behavior -- would crash more quickly .

**Comment by GeorgeSabu** [comment]
@chubukov This bug is there in all versions of maven. This is not related to multiprocessing. This is related to isotopic bar plots. If you disable isotopic barplot this bug does not happen. And also I was not able to get the crash with isotopic barplot enabled and all the isotopic barplot related computation enabled but commented out 3-4 lines which plot the isotopic bar plot.

**Comment by GeorgeSabu** [comment]
@chubukov This is also not a variable of sample size.

**Comment by chubukov** [comment]
@GeorgeSabu totally believe that this is not related to sample size, and that it's related to the isotope bar plot. I'm not sure I'm sold on the idea that it's just drawing the plot -- it seems that the few times we've tracked it through the debugger, the crash has been in various data processing and calculation routines.

Curious to hear how the flag-based calls of the processing routines that you've implemented has affected this behavior.

**Comment by GeorgeSabu** [comment]
@chubukov I replaced the isotopic barplot with an external library. After that, we were not able to reproduce the crash. I was am not sure about the reason why it was happening with the old isotopic plot perfectly. One hypothesis we have is that the number of instances of the barplot that is made is too high because of this mostly stack is overflowing. If we just comment out the part which makes the instances of barplot the crash you reported would not happen.

Flag bases help when to let's say the user keeps on double clicking one of the peaks. The same computation will happen again and again. If a thread is working this thread will be given again the same computation to work on. This was causing the system  to crash on slow systems. But in the current code flag based code is there but not in use. We refactored the peak clicking event listener in a better way and this issue was gone.
